src/metadata.rs

```diff
@@ -0,0 +1,129 @@
+use html5ever::local_name;
+use kuchiki::NodeRef;
+
+
+const TITLE_KEYS: [&str; 6] = [
+    "og:title", "twitter:title", "dc:title", "dcterm:title",
+    "weibo:article:title", "weibo:webpage:title",
+];
+const BYLINE_KEYS: [&str; 6] = [
+    "author", "dc:creator", "dcterm:creator", "og:article:author",
+    "article:author", "byl",
+];
+const DESCRIPTION_KEYS: [&str; 7] = [
+    "description", "dc:description", "dcterm:description", "og:description",
+    "weibo:article:description", "weibo:webpage:description", "twitter:description"
+];
+
+
+pub struct Metadata {
+    pub page_title: Option<String>,
+    pub article_title: Option<String>,
+    pub byline: Option<String>,
+    pub description: Option<String>,
+}
+
+
+pub fn extract(root: &NodeRef) -> Metadata {
+    let mut page_title = root.select_first("title")
+        .map(|node| node.text_contents())
+        .ok();
+
+    let mut article_title = get_article_title(root);
+
+    match (&page_title, &article_title) {
+        (None, Some(at)) => {page_title = Some(at.clone());},
+        (Some(pt), None) => {article_title = Some(pt.clone());},
+        _ => (),
+    }
+
+    let byline = extract_meta_content(root, &BYLINE_KEYS);
+    let description = get_article_description(root);
+    Metadata {page_title, article_title, byline, description}
+}
+
+
+fn get_article_title(root: &NodeRef) -> Option<String> {
+    let meta_title = extract_meta_content(root, &TITLE_KEYS);
+    if meta_title.is_some() {
+        return meta_title;
+    }
+
+    // if no qualifying meta tag is found, look for a single h1
+    // if there are multiple h1s, give up
+    let mut h1s = root.select("h1").unwrap();
```

Comment:
Can this panic? If so you can convert it to an option:

```suggestion
    let mut h1s = root.select("h1").ok();
```

Same for the h2 down there.

Reply:
Ah, you mention that later: "unwrap is safe here because select() only errors when you give it an invalid CSS selector". Nevermind then. Still a nice trick that could come in handy in the future. 😉

Reply:
Sorry, yeah originally that comment appeared the first time I called `unwrap()` in that situation, but then I refactored and forgot to move the comment. :smile:

Reply:
Haha all good. Looking forward to seeing this live. Feel free to merge and publish. I'll do the shuttle part again.